pkg/auth/tokenauthenticator.go

```diff
@@ -38,3 +59,94 @@ func (t *tokenAuthenticator) AddAuthenticationHeader(request *http.Request) erro
 	return true
 }
+
+func IsAuthTypeToken(token string) bool {
+	if _, uuidErr := uuid.Parse(token); uuidErr == nil {
+		return true
+	}
+	return false
+}
+
+func IsAuthTypePAT(token string) bool {
+	// e.g. snyk_uat.1a2b3c4d.mySuperSecret_Token-Value.aChecksum_123-Value
+	patRegex := `^snyk_(?:uat|sat)\.[a-z0-9]{8}\.[a-zA-Z0-9-_]+\.[a-zA-Z0-9-_]+$`
+	if matched, err := regexp.MatchString(patRegex, token); err == nil && matched {
+		return matched
+	}
+	return false
+}
+
+// DeriveEndpointFromPAT makes an authenticated request to the Snyk PAT API and returns the correct Snyk endpoint from the given PAT
+var DeriveEndpointFromPAT = func(token string, config configuration.Configuration, client *http.Client, patServiceUrl string) (string, error) {
+	// TODO: we're implementing a local caching solution here but this will need a proper caching solution
+	// exit early if pat is already cached
+	cached_pat_key := GetCachedPatKeyname(token)
+	cached_pat := config.GetString(cached_pat_key)
+	if len(cached_pat) > 0 {
+		return cached_pat, nil
+	}
+
+	if len(patServiceUrl) == 0 {
+		patServiceUrl = constants.SNYK_DEFAULT_API_URL
+	}
+	apiUrl, err := url.Parse(patServiceUrl)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse Snyk API URL: %w", err)
+	}
+
+	apiUrl.Path = path.Join(apiUrl.Path, patMetadataEndpoint)
+	queryValues := apiUrl.Query()
+	queryValues.Add("version", patAPIVersion)
+	apiUrl.RawQuery = queryValues.Encode()
+	urlString := apiUrl.String()
+
+	req, err := http.NewRequest(http.MethodGet, urlString, nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to create request: %w", err)
+	}
+
+	// add auth headers
+	req.Header.Set("Authorization", fmt.Sprintf("token %s", token))
+
+	res, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		var body any
+		err = json.NewDecoder(res.Body).Decode(&body)
+		if err != nil {
+			return "", fmt.Errorf("failed to get PAT metadata")
+		}
+		return "", fmt.Errorf("failed to get PAT metadata (status: %d): %v", res.StatusCode, body)
+	}
+
+	var patMetadataBody patAPI.GetPATMetadataResponseBody
+	err = json.NewDecoder(res.Body).Decode(&patMetadataBody)
+	if err != nil {
+		return "", fmt.Errorf("failed to decode PAT metadata response: %w", err)
+	}
+
+	if patMetadataBody.Data.Attributes.Hostname == "" {
+		return "", fmt.Errorf("invalid empty hostname")
+	}
+
+	// ensure a match on any of eu.snyk.io, au.snyk.io, us.snyk.io, snyk.io
+	validHostRegex := `^(eu\.snyk\.io|au\.snyk\.io|us\.snyk\.io|snyk\.io)$`
+	if isValid, err := isValidAuthHost(patMetadataBody.Data.Attributes.Hostname, validHostRegex); err != nil || !isValid {
+		return "", fmt.Errorf("invalid hostname: %s", patMetadataBody.Data.Attributes.Hostname)
+	}
+
+	endpoint := fmt.Sprintf("https://api.%s", patMetadataBody.Data.Attributes.Hostname)
+
+	// set the token/endpoint in a cache to avoid infinite loops
+	config.Set(cached_pat_key, endpoint)
```

Comment:
yeah, just seems to be a naming issue with the key and the variable.

Reply:
Hm... this is called from a default function within a configuration.Get(). Do we have potential mutex/locking issues on the configuration here?

Reply:
> Do we have potential mutex/locking issues on the configuration here?
We shouldn't, we have `snyc.RWMutex` implemented in our configuration